### PR TITLE
Config Tweak

### PR DESCRIPTION
**Pull request by nicegamer7**
This tweaks how comments are added to the config files. It also fixes an issue with extra newlines being added in places they shouldn't. See below.

Before:
```
# Multiverse-Inventories Settings


# ===[ Multiverse Inventories Config ]===
settings:

  # This is the locale you wish to use.
  locale: en

  # If this is true it will generate world groups for you based on MV worlds.
  first_run: false

  # If this is set to true, it will enable bypass permissions (Check the wiki for more info.)
  use_bypass: false

  # If set to true, any world not listed in a group will automatically use the settings for the default group!
  default_ungrouped_worlds: false

  # The default and suggested setting for this is FALSE.
  # False means Multiverse-Inventories will not attempt to load or save any player data when they log in and out.
  # That means that MINECRAFT will handle that exact thing JUST LIKE IT DOES NORMALLY.
  # Changing this to TRUE will have Multiverse-Inventories save player data when they log out and load it when they log in.
  # The biggest potential drawback here is that if your server crashes, player stats/inventories may be lost/rolled back!
  save_load_on_log_in_out: false

  # If this is set to true, players will have different inventories/stats for each game mode.
  # Please note that old data migrated to the version that has this feature will have their data copied for both game modes.
  use_game_mode_profiles: false

shares:

  # When set to true, optional shares WILL be utilized in cases where a group does not cover their uses for a world.
  # An example of this in action would be an ungrouped world using last_location.  When this is true, players will return to their last location in that world.
  # When set to false, optional shares WILL NOt be utilized in these cases, effectively disabling it for ungrouped worlds.
  optionals_for_ungrouped_worlds: true

  # You must specify optional shares you wish to use here or they will be ignored.
  # The only built in optional share is "economy"
  use_optionals: []

```
```
# Multiverse-Inventories Groups


# To ADD, DELETE, and EDIT groups use the command /mvinv group.
# No support will be given for those who manually edit these groups.
groups:

  default:

    worlds:

    - world

    - world_the_end

    - world_nether

    shares:

    - ender_chest

    - inventory_contents

    - armor_contents

    - off_hand

    - hit_points

    - remaining_air

    - maximum_air

    - fall_distance

    - fire_ticks

    - xp

    - lvl

    - total_xp

    - food_level

    - exhaustion

    - saturation

    - bed_spawn

    - last_location

    - potion_effects

```

After:
```
# Multiverse-Inventories Settings
settings:
  # This is the locale you wish to use.
  locale: en

  # If this is true it will generate world groups for you based on MV worlds.
  first_run: false

  # If this is set to true, it will enable bypass permissions (Check the wiki for more info.)
  use_bypass: false

  # If set to true, any world not listed in a group will automatically use the settings for the default group!
  default_ungrouped_worlds: false

  # The default and suggested setting for this is FALSE.
  # False means Multiverse-Inventories will not attempt to load or save any player data when they log in and out.
  # That means that MINECRAFT will handle that exact thing JUST LIKE IT DOES NORMALLY.
  # Changing this to TRUE will have Multiverse-Inventories save player data when they log out and load it when they log in.
  # The biggest potential drawback here is that if your server crashes, player stats/inventories may be lost/rolled back!
  save_load_on_log_in_out: false

  # If this is set to true, players will have different inventories/stats for each game mode.
  # Please note that old data migrated to the version that has this feature will have their data copied for both game modes.
  use_game_mode_profiles: false

shares:
  # When set to true, optional shares WILL be utilized in cases where a group does not cover their uses for a world.
  # An example of this in action would be an ungrouped world using last_location.  When this is true, players will return to their last location in that world.
  # When set to false, optional shares WILL NOt be utilized in these cases, effectively disabling it for ungrouped worlds.
  optionals_for_ungrouped_worlds: true

  # You must specify optional shares you wish to use here or they will be ignored.
  # The only built in optional shares are "economy" and "last_location".
  use_optionals: []

```
```
# Multiverse-Inventories Groups
# To ADD, DELETE, and EDIT groups use the command /mvinv group.
# No support will be given for those who manually edit these groups.
groups:
  default:
    worlds:
    - world
    - world_the_end
    - world_nether
    shares:
    - hit_points
    - food_level
    - saturation
    - exhaustion
    - xp
    - total_xp
    - lvl
    - inventory_contents
    - armor_contents
    - bed_spawn
    - maximum_air
    - remaining_air
    - fall_distance
    - fire_ticks
    - potion_effects
    - last_location
    - ender_chest
    - off_hand

```

### CODE REVIEW

**Comment by nicegamer7** [comment]
Sorry, I forgot to note that this change does not cause any ill effects when updating from an older config. It just removes any extra newline characters if there are any.

File: src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java
Comment (dumptruckman) [inline src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java:89]: What if there are line breaks in the header?

File: src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java
Comment (dumptruckman) [inline src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java:212]: Isn't this the actual fix?

**Comment by dumptruckman** [comment]
Very brave of you to modify CommentedYamlConfig. I've always considered it the most hard to refactor code I've ever written.

File: src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java
Comment (nicegamer7) [reply src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java:212]: No, I forget what the actual fix is though. I'll get back to you on this.

File: src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java
Comment (nicegamer7) [reply src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java:89]: I'm unsure, I'll have to test it out and make sure it's handled correctly. Good catch!

File: src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java
Comment (nicegamer7) [reply src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java:212]: The actual issue was the old line 73. The square brackets in the regex caused the resulting array to look something like this:
```
["Line 1...", "", "Line 2...", "", "Line 3", "", etc...
```
Removing the square brackets fixes the issue, then I added Lines 172 and 173 to keep gaps between config options with comments.

File: src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java
Comment (nicegamer7) [inline src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java:81]: I added this to make it so that if a user moves their config files from a system using CRLF to a system using LF (or vice versa), they'd be converted. But now I'm wondering if the server does this automatically. What do you think?

**Comment by nicegamer7** [comment]
It wasn't too bad 😊.

File: src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java
Comment (SlimeDog) [reply src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java:81]: > But now I'm wondering if the server does this automatically. What do you think?

If by "converted" you mean "the server reads them with error" = true. If you mean "files are converted to the type of the system" = false. At least in the DOS->Linux direction (ie., on Linux servers), files are not converted automatically. I convert every DOS-produced file with `dos2unix`. But I do that only for my own convenience and sanity; conversion is not required for the files to be read correctly.

Comment (nicegamer7) [reply src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java:81]: I see... if it's not done automatically, maybe we shouldn't be doing either.

File: src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java
Comment (nicegamer7) [reply src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java:89]: No, it won't be handled correctly, I'll have to make a small change.

File: src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java
Comment (nicegamer7) [reply src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java]: Looking at the source code of the [YamlConfiguration](https://hub.spigotmc.org/stash/projects/SPIGOT/repos/bukkit/browse/src/main/java/org/bukkit/configuration/file/YamlConfiguration.java#119) class, it seems Spigot has chosen to make all config files use the LF line ending, so we will do the the same.

File: src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java
Comment (nicegamer7) [reply src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java]: I fixed it, but now I'm experiencing some strange behavior which I can't understand. I changed the line endings to be LF to match [YamlConfiguration](https://hub.spigotmc.org/stash/projects/SPIGOT/repos/bukkit/browse/src/main/java/org/bukkit/configuration/file/YamlConfiguration.java#119) so that we won't end up making a file with mixed line endings. But if you enter a header with more than one line, the line endings between the lines will be CRLF? I can't figure out where the CR is coming from because it's not YamlConfiguration (the source for the header building method is linked above), and it's not Multiverse-Inventories since I replaced all `System.getProperty("line.separator")` with `\n`... so I'm not sure.

Comment (nicegamer7) [reply src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java]: Just figured it out. It has to do with feildmaster's Configuration-Library. Gonna see what can be done.

Comment (nicegamer7) [reply src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java]: Fixed it by removing Configuration-Library. MV Inventories is a little leaner now 😃.

**Comment by nicegamer7** [comment]
I just rebased this PR onto `main` and added two more commits. The first removes Configuration-Library, which was causing configs with multi-line headers to have mixed line-endings, and the second makes it so that `first_run` is always set to `false` after the first run of the plugin.

With that, this PR is ready for review!

I noticed one final issue, but it's an issue that was already present. If a configuration file cannot be read because it contains invalid YAML, then it is overwritten. I don't think it should be doing that... but I also don't want to fix it in this PR because the diff is beginning to get rather large. I'll open a follow-up PR later on.

**Comment by benwoo1110** [comment]
For the yaml being overwritten issue, I suggest copying the broken file to like a file named `<time>-config.yml.broken`.

Looking at the new config, do you want to retain the somewhat cooler `# ===[ Multiverse Inventories Config ]===` header instead of `# Multiverse-Inventories Settings`>

Also, do you think we should consider sharing this Comments in Config feature and use it in mv-portals and mv-core's configuration file? @nicegamer7 @dumptruckman what do you all think?

Anyways looks good to be me but I think this will need @dumptruckman approval before merging.

**Comment by benwoo1110** [comment]
Oh btw, do you think putting the wiki link to the mv-inv setup guide in the config (probably at the top) will be good? and maybe share list wiki page at the shares part as well.

**Comment by nicegamer7** [comment]
I'm not sure copying the config would be the best idea... since the plugin would keep working, it's possible one might not notice. I think we could disable the plugin. That way, the server owner would be forced to check their console to see what's wrong.

For the header, I don't really like `# ===[ Multiverse Inventories Config ]===` (just personal preference). But I do like the idea of adding the Wiki link.

**Comment by benwoo1110** [comment]
> I'm not sure copying the config would be the best idea... since the plugin would keep working,

The reason for suggesting copying over is because it can cause a config reset on invalid yaml which have reportedly cause people to lose their current config settings.

**Comment by nicegamer7** [comment]
I think if it fails to load, we should just disable the plugin. Although... now that I think about it, can you disable a plugin after the server has already started?

**Comment by SlimeDog** [comment]
> I think if it fails to load, we should just disable the plugin. Although... now that I think about it, can you disable a plugin after the server has already started?

Of course.

File: src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java
Comment (benwoo1110) [inline src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java:68]: we can just do:
```java
try {
    config.save(file);
} catch (Exception e) {
    return false;
}
if (!doComments || comments.isEmpty()) {
    return true;
}
```

File: src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java
Comment (benwoo1110) [inline src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java:71]: Then this wont be needed. One less nesting ;)

File: src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java
Comment (benwoo1110) [inline src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java:125]: Looks like `whiteSpace / 2` done a vary places, may be a good idea to do it once and store it into a variable for reuse.

File: src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java
Comment (benwoo1110) [inline src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java:207]: Use the StringBuilder of `commentString` here too?

**Review by benwoo1110 — CHANGES_REQUESTED** [review]

File: src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java
Comment (nicegamer7) [reply src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java:32]: Yeah I suppose so.

File: src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java
Comment (nicegamer7) [reply src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java:125]: Yup, good idea.

File: src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java
Comment (nicegamer7) [reply src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java:209]: Yeah I ended up doing it a bit differently, but same idea.

File: src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java
Comment (benwoo1110) [inline src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java:166]: Will it be a better idea to have an empty `StringBuilder` here so we use `newLine.append(...)` instead of `newLine.insert(0, ...)` below? iirc append has better time complexity than insert assuming StringBuilder backend logic is similar to arraylist

File: src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java
Comment (nicegamer7) [reply src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java:209]: Actually never mind, I get you meant.

File: src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java
Comment (nicegamer7) [reply src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java:166]: Just searched it up, you're right. Insert's time complexity is worse than append's.

I'm trying some stuff now to reduce the usage of insert.

Comment (benwoo1110) [reply src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java:166]: Yea you should be able to just add the comment lines before the config line, should remove the need for insert for that part.

File: src/test/java/com/onarandombox/multiverseinventories/TestCommentedYamlConfiguration.java
Comment (benwoo1110) [inline src/test/java/com/onarandombox/multiverseinventories/TestCommentedYamlConfiguration.java:19]: Looks like this file generated during the test isn't ignored by git

**Review by dumptruckman — APPROVED** [review]

**Pull request merged**
